### PR TITLE
Add new latency_estimator tracer

### PR DESCRIPTION
**Pull request by hcaseyal**
This latency tracer lets us get a finer-grained latency breakdown of an RPC

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                              FILE SIZE
 ++++++++++++++ GROWING                                ++++++++++++++
  +0.1%    +648 [None]                                 +5.05Ki  +0.1%
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc           +432  +6.2%
       +29%    +266 tcp_do_read                               +266   +29%
       +21%    +178 tcp_flush                                 +178   +21%
  +1.4%    +160 src/core/lib/iomgr/ev_epollex_linux.cc    +160  +1.4%
       +24%    +157 pollable_process_events                   +157   +24%
      +1.0%      +3 [Unmapped]                                  +3  +1.0%
  +0.7%    +112 src/core/lib/surface/call.cc              +112  +0.7%
      +2.2%     +69 call_start_batch                           +69  +2.2%
      +3.5%     +44 finish_batch_step                          +44  +3.5%

  +0.1% +1.32Ki TOTAL                                  +5.74Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +796  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +856  +0.0%



```

File: src/core/lib/iomgr/ev_epollex_linux.cc [hunk 1]
```diff
@@ -767,6 +768,10 @@ static int poll_deadline_to_millis_timeout(grpc_millis millis) {
 }
 
 static void fd_become_readable(grpc_fd* fd, grpc_pollset* notifier) {
+  if (grpc_latency_estimator_trace.enabled()) {
+    gpr_log(GPR_INFO, "latency_estimator:server_inbound_reading:start");
```
Comment (ncteisen) [inline src/core/lib/iomgr/ev_epollex_linux.cc]: don't we need to disambiguate if we are a client or server process? Else how will this interact correctly with the post processor?

File: src/core/lib/iomgr/ev_epollex_linux.cc [hunk 2]
```diff
@@ -818,6 +823,9 @@ static void pollset_shutdown(grpc_pollset* pollset, grpc_closure* closure) {
 static grpc_error* pollable_process_events(grpc_pollset* pollset,
                                            pollable* pollable_obj, bool drain) {
   GPR_TIMER_SCOPE("pollable_process_events", 0);
+  if (grpc_latency_estimator_trace.enabled()) {
+    gpr_log(GPR_INFO, "latency_estimator:server_inbound:start");
```
Comment (ncteisen) [inline src/core/lib/iomgr/ev_epollex_linux.cc]: I don't see a closing server_inbound:end for this one

File: src/core/lib/iomgr/ev_epollex_linux.cc (hunk 1 above)
Comment (hcaseyal) [reply src/core/lib/iomgr/ev_epollex_linux.cc]: I could modify the post processor to have a client and server mode. So when it's in client mode, it throws out all the server-related log statements. 
Maybe the right thing to do at this point is to define where we want to put the start and end markers for each leg of an rpc? And perhaps that will solve the disambiguation problem just by nature of where the markers are.

Comment (hcaseyal) [reply src/core/lib/iomgr/ev_epollex_linux.cc]: @ncteisen Talked with Abhishek, and he said it's ok to leave the log statements where they are now, as long as it's consistent between runs. I think modifying the post processor to have the two separate modes should work as a quick solution.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +760 [None]                                  +6.55Ki  +0.1%
      +0.1%    +696 [Unmapped]                              +6.55Ki  +0.1%
      +3.7%     +40 [None]                                        0  [ = ]
      [NEW]     +24 grpc_latency_estimator_trace                  0  [ = ]
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc            +432  +6.2%
       +29%    +266 tcp_do_read                                +266   +29%
       +21%    +178 tcp_flush                                  +178   +21%
  +0.7%    +112 src/core/lib/surface/call.cc               +112  +0.7%
      +2.2%     +69 call_start_batch                            +69  +2.2%
      +3.5%     +44 finish_batch_step                           +44  +3.5%
  +0.7%     +80 src/core/lib/iomgr/ev_epollex_linux.cc      +80  +0.7%
       +13%     +88 pollable_process_events                     +88   +13%
  [NEW]     +48 src/core/lib/debug/latency_estimator.cc     +48  [NEW]
      [NEW]     +48 _GLOBAL__sub_I_latency_estimator.cc         +48  [NEW]

  +0.1% +1.40Ki TOTAL                                   +7.20Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +796  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +856  +0.0%



```

File: src/core/lib/iomgr/ev_epollex_linux.cc (hunk 1 above)
Comment (ncteisen) [reply src/core/lib/iomgr/ev_epollex_linux.cc]: That solution sounds good to me!

**Comment by ncteisen** [comment]
Looks like some sanity checks are failing

run `tools/distrib/clang_format_code.sh` (that will autofix)

Then the other two failing ones need manual changes, but they should ideally fail in ways that make it clear how to fix. let me know if that is not the case.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +760 [None]                                  +6.55Ki  +0.1%
      +0.1%    +696 [Unmapped]                              +6.55Ki  +0.1%
      +3.7%     +40 [None]                                        0  [ = ]
      [NEW]     +24 grpc_latency_estimator_trace                  0  [ = ]
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc            +432  +6.2%
       +29%    +266 tcp_do_read                                +266   +29%
       +21%    +178 tcp_flush                                  +178   +21%
  +0.7%    +112 src/core/lib/surface/call.cc               +112  +0.7%
      +2.2%     +69 call_start_batch                            +69  +2.2%
      +3.5%     +44 finish_batch_step                           +44  +3.5%
  +0.7%     +80 src/core/lib/iomgr/ev_epollex_linux.cc      +80  +0.7%
       +13%     +88 pollable_process_events                     +88   +13%
  [NEW]     +48 src/core/lib/debug/latency_estimator.cc     +48  [NEW]
      [NEW]     +48 _GLOBAL__sub_I_latency_estimator.cc         +48  [NEW]

  +0.1% +1.40Ki TOTAL                                   +7.20Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +844  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +904  +0.0%



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +760 [None]                                  +6.55Ki  +0.1%
      +0.1%    +696 [Unmapped]                              +6.55Ki  +0.1%
      +3.7%     +40 [None]                                        0  [ = ]
      [NEW]     +24 grpc_latency_estimator_trace                  0  [ = ]
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc            +432  +6.2%
       +29%    +266 tcp_do_read                                +266   +29%
       +21%    +178 tcp_flush                                  +178   +21%
  +0.7%    +112 src/core/lib/surface/call.cc               +112  +0.7%
      +2.2%     +69 call_start_batch                            +69  +2.2%
      +3.5%     +44 finish_batch_step                           +44  +3.5%
  +0.7%     +80 src/core/lib/iomgr/ev_epollex_linux.cc      +80  +0.7%
       +13%     +88 pollable_process_events                     +88   +13%
  [NEW]     +48 src/core/lib/debug/latency_estimator.cc     +48  [NEW]
      [NEW]     +48 _GLOBAL__sub_I_latency_estimator.cc         +48  [NEW]

  +0.1% +1.40Ki TOTAL                                   +7.20Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +844  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +904  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +760 [None]                                  +6.55Ki  +0.1%
      +0.1%    +696 [Unmapped]                              +6.55Ki  +0.1%
      +3.7%     +40 [None]                                        0  [ = ]
      [NEW]     +24 grpc_latency_estimator_trace                  0  [ = ]
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc            +432  +6.2%
       +29%    +266 tcp_do_read                                +266   +29%
       +21%    +178 tcp_flush                                  +178   +21%
  +0.7%    +112 src/core/lib/surface/call.cc               +112  +0.7%
      +2.2%     +69 call_start_batch                            +69  +2.2%
      +3.5%     +44 finish_batch_step                           +44  +3.5%
  +0.7%     +80 src/core/lib/iomgr/ev_epollex_linux.cc      +80  +0.7%
       +13%     +88 pollable_process_events                     +88   +13%
  [NEW]     +48 src/core/lib/debug/latency_estimator.cc     +48  [NEW]
      [NEW]     +48 _GLOBAL__sub_I_latency_estimator.cc         +48  [NEW]

  +0.1% +1.40Ki TOTAL                                   +7.20Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +844  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +904  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +760 [None]                                  +6.55Ki  +0.1%
      +0.1%    +696 [Unmapped]                              +6.55Ki  +0.1%
      +3.7%     +40 [None]                                        0  [ = ]
      [NEW]     +24 grpc_latency_estimator_trace                  0  [ = ]
  +6.2%    +432 src/core/lib/iomgr/tcp_posix.cc            +432  +6.2%
       +29%    +266 tcp_do_read                                +266   +29%
       +21%    +178 tcp_flush                                  +178   +21%
  +0.7%    +112 src/core/lib/surface/call.cc               +112  +0.7%
      +2.2%     +69 call_start_batch                            +69  +2.2%
      +3.5%     +44 finish_batch_step                           +44  +3.5%
  +0.7%     +80 src/core/lib/iomgr/ev_epollex_linux.cc      +80  +0.7%
       +13%     +88 pollable_process_events                     +88   +13%
  [NEW]     +48 src/core/lib/debug/latency_estimator.cc     +48  [NEW]
      [NEW]     +48 _GLOBAL__sub_I_latency_estimator.cc         +48  [NEW]

  +0.1% +1.40Ki TOTAL                                   +7.20Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.1%    +116 [None]                                                                                  +844  +0.0%
  +1.3%     +60 src/cpp/client/channel_cc.cc                                                             +60  +1.3%
      [NEW]    +194 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +194  [NEW]
      +4.7%     +60 grpc::Channel::CreateCall                                                                +60  +4.7%

  +0.0%    +176 TOTAL                                                                                   +904  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
https://github.com/grpc/grpc/blob/18ce0cf26e733647b33cbaafee57245fdfb932e7/src/core/lib/surface/call.cc#L1303 is a better place for server inbound proc end.

**Comment by yashykt** [comment]
https://github.com/grpc/grpc/blob/18ce0cf26e733647b33cbaafee57245fdfb932e7/src/core/lib/surface/call.cc#L1602 for server outbound start

**Comment by yashykt** [comment]
I'll leave it up to @ncteisen  as to whether this trace flag should be a compile time thing.

**Pull request closed without merging**